src/std/misc/string.ss

```diff
@@ -77,3 +79,80 @@ package: std/misc
     ((_ eol fallback) (let ((trimmed (string-trim-suffix eol string)))
                         (if (eq? trimmed string) fallback (values trimmed eol)))))
   (try +crlf+ (try +lf+ (try +cr+ (values string "")))))
+
+
+;; Returns a single character-string of s at the index i
+;; or an error if the index could not be found.
+(def (string-at s i)
+  (string (string-ref s i)))
+
+
+;; string-replace helper which handels the case
+;; that the argument 'old' is an empty string.
+;;   new    non-empty
+;;   limit  non-zero
+(def (sr-helper-empty-old str new limit)
+  (def len-str (string-length str))
+  (if (= limit 1)
+    (string-append new str)
+    (let ((stop (1- (string-length str)))
+	  (limit (if (or (< limit 0)
+			 (> limit len-str)) (1+ len-str) limit)))
+      (string-concatenate
+       (cons new
+	(let loop ((i 0) (count 1))
+	  (cond
+	   ((= count limit)
+	    (cons new (cons (substring/shared str i) [])))
+	   ((= i stop)
+	    (cons (if (zero? i) "" new)
+		  (cons (string-at str i) (cons new []))))
+	   (else
+	    (cons (if (zero? i) "" new)
+		  (cons (string-at str i)
+			(loop (1+ i) (1+ count))))))))))))
+
+
+;; string-replace helper which handels the case
+;; that the argument 'old' is a non-empty string.
+;; str    non-empty
+;; old    non-empty
+;; new    can be empty
+;; limit  non-zero
+(def (sr-helper-nonempty-old str old new limit)
+  (def len-str (string-length str))
+  (def len-old (string-length old))
+  (def stop (1- (string-length str)))
+  (string-concatenate
+   (let loop ((i 0) (count 0) (last 0) (j 0))
+     (cond
+      ((= count limit)
+       (cons (substring/shared str i) []))
+      ((= i stop)
+       (if (and (eq? (string-ref str i) (string-ref old j))
+		(= j (1- len-old)))
+	 (cons new [])
+	 (cons (substring/shared str last) [])))
+      (else
+       (if (eq? (string-ref str i) (string-ref old j))
+	 (if (= j (1- len-old))
+	   (cons new (loop (1+ i) (1+ count) (1+ i) 0))
+	   (loop (1+ i) count last (1+ j)))
+	 (cons (substring/shared str last (1+ i))
+	       (loop (1+ i) count (1+ i) 0))))))))
+
+;; In the copy of str replace the string old with string new.
+;; Limit controls the number of replacements (limit < 0 equals to #f).
+(def (string-substitute str old new (limit #f))
+  (let ((limit (if (and (number? limit) (> limit -1)) limit -1))
+	(old-empty? (string-empty? old))
+	(new-empty? (string-empty? new))
+	(str-empty? (string-empty? str)))
+    (cond
+     ((or (zero? limit)
+	  (and old-empty? new-empty?)) str)
+     (old-empty?
+      (sr-helper-empty-old str new limit))
```

Comment:
what are the semantics of this? You shouldn't be replacing anything if the old string is empty.

Reply:
An empty old string means replace always, including before the first character and after the last character. This is how other languages handle it, too. Therefor I use two different strategies to resolve the task.

```
(string-replace "abc" "" "_") => "_a_b_c_"
```

Reply:
ok, that's reasonable.